lib/open_api_spex/cast.ex

```diff
@@ -207,6 +207,10 @@ defmodule OpenApiSpex.Cast do
   def cast(%__MODULE__{schema: %{type: _other}} = ctx),
     do: error(ctx, {:invalid_schema_type})
 
+  def cast(%__MODULE__{value: struct, schema: schema_module} = ctx)
+      when is_struct(struct) and is_atom(schema_module),
+      do: cast(ctx.schema.schema(), ctx.value)
```

Comment:
We avoid calling the `schema()` function like this during casting as it can have performance impacts.
The full API schema should be resolved before casting using [OpenApiSpex.resolve_schema_modules](https://hexdocs.pm/open_api_spex/OpenApiSpex.html#resolve_schema_modules/1)

Once resolved the `schema_module` should be replaced by a `%Reference{}` struct which will be handled by an earlier clause.